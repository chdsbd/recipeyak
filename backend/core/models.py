import hashlib
from typing import List, Union
import logging
import itertools

from django.db import models, transaction
from django.contrib.auth.models import AbstractBaseUser, BaseUserManager, PermissionsMixin
from django.contrib.contenttypes.fields import GenericForeignKey, GenericRelation
from django.contrib.contenttypes.models import ContentType
from django.contrib.postgres.fields import CIEmailField
from django.core.validators import MinValueValidator

from allauth.socialaccount.models import EmailAddress

logger = logging.getLogger(__name__)


class MyUserManager(BaseUserManager):
    def create_user(self, email, password=None):
        """
        Creates and saves a user with given email and password.
        """
        if not email:
            raise ValueError('User must have an email address')
        with transaction.atomic():
            user = self.model(email=self.normalize_email(email))
            user.set_password(password)
            user.save(using=self._db)
            EmailAddress.objects.create(user=user, email=email)
            logger.info(f'Created new user: {user}')
            return user

    def create_superuser(self, email, password):
        """
        Creates and saves a superuser with the given email and password.
        """
        user = self.create_user(email=email, password=password)
        user.is_admin = True
        user.is_superuser = True
        user.save(using=self._db)
        return user


class MyUser(AbstractBaseUser, PermissionsMixin):
    """Custom user model that only requires an email and password"""
    email = CIEmailField(unique=True)

    # required for admin
    is_active = models.BooleanField(default=True)
    is_admin = models.BooleanField(default=False)

    created = models.DateField(auto_now_add=True)
    last_updated = models.DateField(auto_now=True)

    recipes = GenericRelation('Recipe', related_query_name='owner_user')

    USERNAME_FIELD = 'email'
    REQUIRED_FIELDS: List[str] = []

    objects = MyUserManager()

    def get_full_name(self):
        return self.email

    def get_short_name(self):
        return self.email

    # required for admin
    def has_perm(self, perm, obj=None):
        """Does the user have a specific permission?"""
        # TODO: Add permissions
        return True

    # required for admin
    def has_module_perms(self, app_label):
        """Does the user have permissions to view the app `app_label`?"""
        # TODO: Add permissions
        return True

    def has_invite(self, team) -> bool:
        """
        Return if user has invite to team.
        """
        return self.membership_set.filter(team=team).exclude(invite=None).exists()  # type: ignore

    def has_team(self):
        return self.membership_set.filter(invite=None).count() > 0

    # required for admin
    @property
    def is_staff(self):
        return self.is_admin

    @property
    def avatar_url(self):
        md5_email = hashlib.md5(self.email.encode('utf-8')).hexdigest()
        # indenticons by default `d=identicon`
        # Avatars with ratings of G only `r=g`
        # https://secure.gravatar.com/site/implement/images/
        return f'//www.gravatar.com/avatar/{md5_email}?d=identicon&r=g'

    @property
    def scheduled_recipes(self):
        return ScheduledRecipe.objects.filter(user=self)

    def __str__(self):
        return self.email


class CommonInfo(models.Model):
    """Abstract model for storing common model info"""
    created = models.DateTimeField(auto_now_add=True)
    modified = models.DateTimeField(auto_now=True)

    class Meta:
        abstract = True


class Recipe(CommonInfo):
    name = models.CharField(max_length=255)
    author = models.CharField(max_length=255, blank=True, null=True)
    source = models.CharField(max_length=255, blank=True, null=True)
    time = models.CharField(max_length=255, blank=True, null=True)
    servings = models.CharField(max_length=255, blank=True, null=True)

    edits = models.IntegerField(default=0, editable=False)

    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
    object_id = models.PositiveIntegerField()
    owner = GenericForeignKey('content_type', 'object_id')

    def move_to(self, account):
        """
        Move recipe from current owner to another team or user

        All we need to do is change the owner for this one.
            - Steps and Ingredients will be fine since we aren't changing pk's
        """
        with transaction.atomic():
            self.owner = account
            self.save()
            return self

    def copy_to(self, account):
        """
        Copy recipe to another team or user
        """
        with transaction.atomic():
            # clone top level recipe object
            recipe_copy = Recipe.objects.get(pk=self.pk)
            recipe_copy.pk = None
            recipe_copy.owner = account
            recipe_copy.save()

            for recipe_component in itertools.chain(self.steps, self.ingredients):
                recipe_component.pk = None
                # alternative to recipe_copy.steps_set.add(step)
                # this is required due to the db constraint on recipe, position
                recipe_component.recipe_id = recipe_copy.pk
                recipe_component.save()
            recipe_copy.save()
            return recipe_copy

    def schedule(self, on, user, count):
        return ScheduledRecipe.objects.create_scheduled(recipe=self, on=on, count=count, user=user)

    @property
    def ingredients(self):
        """Return recipe ingredients ordered by creation date"""
        return Ingredient.objects.filter(recipe=self).order_by('created')

    @property
    def steps(self):
        """Return recipe steps ordered by creation date"""
        return Step.objects.filter(recipe=self).order_by('position', 'created')

    @property
    def tags(self):
        """Return recipe tags ordered by creation date"""
        return Tag.objects.filter(recipe=self).order_by('created')

    def __str__(self):
        return f'{self.name} by {self.author}'

    def save(self, *args, **kwargs):
        is_new = self.pk is None
        if not is_new:
            # we only want to increment the edits if we aren't setting the
            # edits field specifically
            edits_unchanged = Recipe.objects.get(pk=self.id).edits == self.edits
            if edits_unchanged:
                self.edits += 1
        super().save(*args, **kwargs)


class Ingredient(CommonInfo):
    """
    Recipe ingredient

    ex:
        1 medium tomato, diced

    quantity = 1 medium
    name = tomato
    description = diced

    """
    quantity = models.CharField(max_length=255)
    name = models.CharField(max_length=255)
    description = models.CharField(max_length=255, blank=True)
    recipe = models.ForeignKey(Recipe, on_delete=models.CASCADE)
    position = models.FloatField()

    class Meta:
        unique_together = (('recipe', 'position'),)
        ordering = ['-position', ]

    def __str__(self):
        return f'{self.quantity} {self.name} {self.description}'

    def __repr__(self):
        return f'<quantity={self.quantity} {self.name} description={self.description} recipe={self.recipe}>'


class Step(CommonInfo):
    """Recipe step"""
    text = models.TextField()
    recipe = models.ForeignKey(Recipe, on_delete=models.CASCADE)
    position = models.FloatField()

    class Meta:
        unique_together = (('recipe', 'position'),)
        ordering = ['-position', ]

    def __str__(self):
        return self.text


class Tag(CommonInfo):
    """Recipe tag"""
    text = models.CharField(max_length=255)
    recipe = models.ForeignKey(Recipe, on_delete=models.CASCADE)

    def __str__(self):
        return self.text


class ScheduledRecipeManager(models.Manager):
    def create_scheduled(self, recipe: Recipe, on, count: int, user: MyUser) -> 'ScheduledRecipe':
        """
        add to existing scheduled recipe count for dupes
        """
        with transaction.atomic():
            existing = ScheduledRecipe.objects.filter(recipe=recipe, on=on, user=user).first()
            if existing:
                existing.count += count
                existing.save()
                return existing
            else:
                return ScheduledRecipe.objects.create(recipe=recipe, on=on, count=count, user=user)


class ScheduledRecipe(CommonInfo):
    recipe = models.ForeignKey(Recipe, on_delete=models.CASCADE)
    on = models.DateField()
    count = models.PositiveIntegerField(validators=[MinValueValidator(1)])
    user = models.ForeignKey(MyUser, on_delete=models.CASCADE)

    objects = ScheduledRecipeManager()

    class Meta:
        unique_together = (('recipe', 'on', 'user'),)

    def __str__(self):
        return f'ScheduledRecipe:: {self.count} of {self.recipe.name} on {self.on} for {self.user}'


class InviteManager(models.Manager):
    def create_invite(self, email, team, level, creator) -> 'Invite':
        user = MyUser.objects.filter(email=email).first()
        if not user:
            user = MyUser.objects.create_user(email=email)
        m = Membership.objects.create(user=user, team=team, level=level, is_active=False)
        invite: Invite = self.model.objects.create(membership=m, creator=creator)
        return invite


class Invite(CommonInfo):
    membership = models.OneToOneField('Membership', on_delete=models.CASCADE)
    creator = models.ForeignKey(MyUser, on_delete=models.CASCADE)

    OPEN = 'open'
    DECLINED = 'declined'
    ACCEPTED = 'accepted'

    INVITE_STATUS = (
        (OPEN, OPEN),
        (DECLINED, DECLINED),
        (ACCEPTED, ACCEPTED),
    )

    status = models.CharField(
        max_length=11,
        choices=INVITE_STATUS,
        default=OPEN,
    )

    objects = InviteManager()

    def __str__(self):
        return f'<Invite â€¢ Membership: {self.membership}>'

    @property
    def user(self):
        return self.membership.user

    @property
    def active(self):
        return self.membership.is_active

    @property
    def team(self):
        return self.membership.team

    def accept(self):
        self.membership.set_active()
        self.status = self.ACCEPTED
        self.save()

    def decline(self):
        self.status = self.DECLINED
        self.save()


class Team(CommonInfo):
    name = models.CharField(max_length=255)
    is_public = models.BooleanField(default=False)
    recipes = GenericRelation('Recipe', related_query_name='owner_team')

    def __str__(self):
        return f'<Team â€¢ name: {self.name}, is_public: {self.is_public}>'

    def force_join(self, user, level=None):
        with transaction.atomic():
            if level is None:
                level = Membership.CONTRIBUTOR
            m, created = Membership.objects.get_or_create(team=self, user=user, defaults={'level': level, 'is_active': True})
            if not created:
                m.level = level
                m.is_active = True
                m.save()
            # remove existing invite
            if user.has_invite(self):
                user.membership_set.exclude(invite=None).get(team=self).invite.delete()
            return m

    def force_join_admin(self, user):
        return self.force_join(user, level=Membership.ADMIN)

    def invite_user(self, user, creator, level=None) -> Invite:
        """
        Invite user to team

        Adds member to team with is_active=False. Creates associated Invite.
        """
        if level is None:
            level = Membership.CONTRIBUTOR
        return Invite.objects.create_invite(email=user.email,
                                            team=self,
                                            level=level,
                                            creator=creator)

    def kick_user(self, user):
        """
        Remove user from team. If they have an invite, remove it as well.
        """
        membership = user.membership_set.get(team=self)
        # delete membership. By deleting, associated invites will be deleted.
        membership.delete()

    def set_public(self):
        self.is_public = True
        self.save()

    def set_private(self):
        self.is_public = False
        self.save()

    def admins(self):
        return self.membership_set.filter(is_active=True, level=Membership.ADMIN)

    def is_member(self, user) -> bool:
        return self.membership_set.filter(user=user, is_active=True).exists()  # type: ignore

    def is_contributor(self, user) -> bool:
        return self.membership_set.filter(user=user, is_active=True, level=Membership.CONTRIBUTOR).exists()  # type: ignore

    def is_admin(self, user) -> bool:
        return self.membership_set.filter(user=user, is_active=True, level=Membership.ADMIN).exists()  # type: ignore

    def invite_exists(self, email: Union[MyUser, str]) -> bool:
        return Membership.objects.filter(team=self, user__email=email).exists()  # type: ignore


class Membership(CommonInfo):
    ADMIN = 'admin'
    CONTRIBUTOR = 'contributor'
    READ_ONLY = 'read'

    MEMBERSHIP_CHOICES = (
        (ADMIN, ADMIN),
        (CONTRIBUTOR, CONTRIBUTOR),
        (READ_ONLY, READ_ONLY),
    )

    level = models.CharField(
        max_length=11,
        choices=MEMBERSHIP_CHOICES,
        default=CONTRIBUTOR,
    )

    team = models.ForeignKey(Team, on_delete=models.CASCADE)
    user = models.ForeignKey(MyUser, on_delete=models.CASCADE)

    class Meta:
        unique_together = (('user', 'team'),)

    # A user is activated once they accept their invite
    is_active = models.BooleanField(default=False)

    def set_active(self):
        self.is_active = True
        self.save()

    def save(self, *args, **kwargs):
        is_new = self.pk is None
        if not is_new:
            # NOTE: although we check inside the serializer to prevent demoting the
            # last admin, this forms a last line of defence
            current = Membership.objects.get(pk=self.pk)
            one_admin_left = len(self.team.admins()) == 1
            demoting_admin = current.level == self.ADMIN and self.level != self.ADMIN
            if one_admin_left and demoting_admin:
                raise ValueError('cannot demote self as last admin')
        super().save(*args, **kwargs)

    def delete(self):
        last_member = self.team.membership_set.count() == 1
        if last_member:
            raise ValueError('cannot delete last member of team')
        super().delete()

    def __str__(self):
        return f'<Membership â€¢ user_email: {self.user.email}, team: {self.team.id} level: {self.level}>'
