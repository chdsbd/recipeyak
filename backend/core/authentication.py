from rest_framework.authentication import (
    SessionAuthentication as DRFSessionAuthentication,
)


class SessionAuthentication(DRFSessionAuthentication):
    def enforce_csrf(self, request):
        """
        We don't use CSRF tokens to prevent CSRF attacks. Instead we use
        SameSite: Lax in our cookies.

        see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite
        """
        return
