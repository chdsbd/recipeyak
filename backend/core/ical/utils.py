from datetime import date, datetime
from typing import Sequence, cast

from icalendar import Calendar, Event, vDate, vDatetime


def create_event(
    *,
    id: str,
    name: str,
    description: str,
    url: str,
    start_date: date,
    end_date: date,
    created: datetime,
) -> Event:
    event = Event(
        uid=id,
        dtstart=vDate(start_date),
        dtend=vDate(end_date),
        summary=name,
        description=description,
        url=url,
        transp="TRANSPARENT",  # if it is a date, then we use TRANSPARENT, else OPAQUE
        created=vDatetime(created),
        dtstamp=vDatetime(created),
    )
    event["LAST-MODIFIED"] = vDatetime(created)
    return event


def create_calendar(
    *, name: str, description: str, events: Sequence[Event]
) -> Calendar:
    cal = Calendar(
        prodid="-//Recipe Yak//Schedule//EN", calscale="GREGORIAN", version=2.0
    )
    cal["x-wr-calname"] = name
    cal["x-wr-caldesc"] = description
    for event in events:
        cal.add_component(event)
    return cal


def to_ical_time(date: datetime) -> str:
    return cast(str, vDatetime(date).to_ical().decode())
