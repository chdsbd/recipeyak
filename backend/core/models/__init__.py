from django.db.models import Q

from core.models.ingredient import Ingredient  # noqa: F401
from core.models.invite import Invite  # noqa: F401
from core.models.membership import Membership, get_random_ical_id  # noqa: F401
from core.models.note import Note  # noqa: F401
from core.models.recipe import Recipe  # noqa: F401
from core.models.recipe_change import ChangeType, RecipeChange  # noqa: F401
from core.models.scheduled_recipe import ScheduledRecipe  # noqa: F401
from core.models.section import Section  # noqa: F401
from core.models.shopping_list import ShoppingList  # noqa: F401
from core.models.step import Step  # noqa: F401
from core.models.team import Team  # noqa: F401
from core.models.timeline_event import TimelineEvent  # noqa: F401
from core.models.user import User  # noqa: F401
from core.models.upload import Upload  # noqa: F401
from core.models.note_upload import NoteUpload  # noqa: F401


def user_active_team_ids(user):
    return user.membership_set.filter(is_active=True).values_list("team")


def user_and_team_recipes(user: User):
    return Recipe.objects.filter(
        Q(owner_user=user) | Q(owner_team__in=user_active_team_ids(user))
    )
