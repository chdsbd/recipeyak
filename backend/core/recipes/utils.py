def add_positions(recipe_steps):
    """Add `position` to step data if missing."""
    missing_position = any(s.get("position") is None for s in recipe_steps)
    if missing_position:
        for i, step in enumerate(recipe_steps):
            step["position"] = i + 10.0
    return recipe_steps
