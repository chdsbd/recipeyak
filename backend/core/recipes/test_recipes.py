from datetime import datetime

import pytest
from django.utils.dateparse import parse_datetime
from rest_framework import status
from rest_framework.test import APIClient
from syrupy.matchers import path_type
from core import ordering

from core.models import (
    Ingredient,
    Membership,
    Note,
    Recipe,
    Step,
    Team,
    TimelineEvent,
    User,
)

pytestmark = pytest.mark.django_db


def test_recipe_creation(client, user, team):
    """
    ensure that the user can create recipes
    """
    client.force_authenticate(user)

    data = {
        "team": team.id,
        "name": "Recipe name",
    }

    res = client.post("/api/v1/recipes/", data)
    assert res.status_code == status.HTTP_201_CREATED

    recipe_id = res.json().get("id")
    res = client.get(f"/api/v1/recipes/{recipe_id}/")
    assert res.status_code == status.HTTP_200_OK

    assert isinstance(
        parse_datetime(res.json()["created"]), datetime
    ), "return a valid date time string"


def test_creating_recipe_with_empty_ingredients_and_steps(client, user, team):
    """
    ensure we can create a recipe without ingredients or steps
    """

    client.force_authenticate(user)

    data = {"name": "some recipe", "team": team.id}

    res = client.post("/api/v1/recipes/", data)
    assert res.status_code == 201


def test_cache_headers(client: APIClient, user: User, recipe: Recipe) -> None:
    """
    check that we actually disable caching
    """
    client.force_authenticate(user)

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK
    assert res["Cache-Control"] == "no-store, no-cache, must-revalidate"


def test_recipe_creation_for_a_team(client, team, user):
    """
    ensure that the user can create recipe for a team
    """

    assert team.is_member(user)

    client.force_authenticate(user)

    assert TimelineEvent.objects.count() == 0

    data = {
        "name": "Recipe name",
        "ingredients": [
            {
                "quantity": "1",
                "unit": "tablespoon",
                "name": "black pepper",
                "description": "",
            }
        ],
        "steps": [{"text": "place fish in salt"}],
        "team": team.id,
    }

    url = "/api/v1/recipes/"

    res = client.post(url, data)
    assert res.status_code == status.HTTP_201_CREATED

    assert (
        TimelineEvent.objects.count()
        == TimelineEvent.objects.filter(action="created").count()
        == 1
    )

    recipe_id = res.json()["id"]

    assert isinstance(Recipe.objects.get(id=recipe_id).owner, Team)

    assert client.get(f"/api/v1/recipes/{recipe_id}/").status_code == status.HTTP_200_OK

    assert Team.objects.get(id=team.id).recipes.first().id == recipe_id


def test_recipe_deletion(client, user, recipe):
    """
    ensure that the user can delete a recipe
    """
    client.force_authenticate(user)

    res = client.delete(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_204_NO_CONTENT

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_404_NOT_FOUND


def test_recipe_updating(client, user, recipe):
    """
    ensure a user can update a recipe
    """
    client.force_authenticate(user)

    data = {"name": "A new name"}

    assert recipe.name != data.get("name")

    res = client.patch(f"/api/v1/recipes/{recipe.id}/", data)
    assert res.status_code == status.HTTP_200_OK

    assert res.json().get("name") == data.get("name")


def test_recipe_archived_at(client, user, recipe):
    """
    check we can set archived_at on the recipe
    """
    client.force_authenticate(user)

    res = client.get("/api/v1/recipes/")
    assert res.status_code == status.HTTP_200_OK
    assert len(res.json()) == 1
    assert res.json()[0]["archived_at"] is None, "archived_at should be on list view"

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK
    assert res.json()["archived_at"] is None, "archived_at should be on detail view"

    assert TimelineEvent.objects.count() == 0

    data = {"archived_at": datetime.now()}
    res = client.patch(f"/api/v1/recipes/{recipe.id}/", data)
    assert res.status_code == status.HTTP_200_OK
    assert isinstance(res.json()["archived_at"], str), "should be a date string"

    res = client.get("/api/v1/recipes/")
    assert res.status_code == status.HTTP_200_OK
    assert len(res.json()) == 1
    assert isinstance(res.json()[0]["archived_at"], str), "should be a date string"
    assert all(
        item["created_by"]["email"] == user.email
        for item in res.json()[0]["timelineItems"]
    ), "should have user email"

    assert (
        TimelineEvent.objects.count()
        == TimelineEvent.objects.filter(action="archived").count()
        == 1
    )

    res = client.patch(f"/api/v1/recipes/{recipe.id}/", {"archived_at": None})
    assert res.status_code == status.HTTP_200_OK
    assert res.json()["archived_at"] is None

    assert TimelineEvent.objects.count() == 2
    assert TimelineEvent.objects.filter(action="archived").count() == 1
    assert TimelineEvent.objects.filter(action="unarchived").count() == 1


def test_recipe_list_timeline_items(client, user, recipe, snapshot):
    """
    Verify we list timeline items with the correct keys.
    """
    client.force_authenticate(user)

    res = client.get("/api/v1/recipes/")
    assert res.status_code == status.HTTP_200_OK
    assert len(res.json()) == 1
    timeline_items = res.json()[0]["timelineItems"]
    assert sorted(timeline_items, key=lambda x: str(x["text"])) == snapshot(
        matcher=path_type(
            {
                r".*\.recipe_id": (int,),
                r".*\.created": (str,),
                r".*\.modified": (str,),
                r".*\.id": (int,),
                r".*\.last_modified_by\.id": (int,),
                r".*\.created_by\.id": (int,),
            },
            regex=True,
        )
    )


def test_updating_step_of_recipe(client, user, recipe):
    """
    ensure a user can update an step of a recipe
    """
    client.force_authenticate(user)

    step = recipe.steps[0]

    step_data = {
        "text": "An updated step",
        "position": step.position + chr(ordering.END_CHAR_CODE),
    }

    url = f"/api/v1/recipes/{recipe.id}/steps/{step.id}/"

    res = client.patch(url, step_data)
    assert res.status_code == status.HTTP_200_OK

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK
    updated_step = next((x for x in res.json()["steps"] if x["id"] == step.pk))

    assert updated_step.get("id") is not None

    for key, value in step_data.items():
        assert updated_step.get(key) == value


def test_deleting_step_from_recipe(client, user, recipe):
    """
    ensure a user can remove a step from a recipe
    """
    client.force_authenticate(user)

    step_id = recipe.steps[0].id

    assert isinstance(step_id, int)

    res = client.delete(f"/api/v1/recipes/{recipe.id}/steps/{step_id}/")
    assert res.status_code == status.HTTP_204_NO_CONTENT

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK

    assert step_id not in (
        step.get("id") for step in res.json().get("steps")
    ), "step was still in the recipe after being deleted"


ingredient = {
    "quantity": "1",
    "unit": "tablespoon",
    "name": "black pepper",
    "description": "",
}
step = {"text": "place fish in salt"}


@pytest.mark.parametrize(
    "data,url_template,model",
    [
        (ingredient, "/api/v1/recipes/{recipe_id}/ingredients/", Ingredient),
        (step, "/api/v1/recipes/{recipe_id}/steps/", Step),
    ],
)
def test_position_step_ingredient(client, user, recipe, data, url_template, model):
    """
    Catch bug in positioning code where soft deletion would hide already taken
    position from application code. This triggered an integrity error at the
    database level.
    """
    client.force_authenticate(user)
    url = url_template.format(recipe_id=recipe.id)

    res = client.post(url, data)
    assert res.status_code == status.HTTP_201_CREATED
    model.objects.get(id=res.json()["id"]).delete()
    res = client.post(url, data)
    assert res.status_code == status.HTTP_201_CREATED


def test_adding_note_to_recipe(client, user, recipe):
    Note.objects.all().delete()
    client.force_authenticate(user)
    data = {"text": "use a mixer to speed things along.", "attachment_upload_ids": []}
    res = client.post(f"/api/v1/recipes/{recipe.id}/notes/", data)
    assert status.is_success(res.status_code)
    assert res.json()["created_by"]["id"] == user.id
    assert res.json()["text"] == data["text"]


def test_modifying_note_of_recipe(client, user, user2, recipe):
    note = recipe.notes.first()
    client.force_authenticate(user)
    data = {"text": "preheat the oven!!"}
    res = client.patch(f"/api/v1/notes/{note.id}/", data)
    assert status.is_success(res.status_code)
    assert res.json()["last_modified_by"]["id"] == user.id
    assert res.json()["created_by"]["id"] == user.id


def test_delete_note(client, user, user2, recipe):
    note = recipe.notes.first()
    client.force_authenticate(user)
    before_count = recipe.notes.count()
    assert before_count > 0
    res = client.delete(f"/api/v1/notes/{note.id}/")
    assert status.is_success(res.status_code)
    assert recipe.notes.count() == before_count - 1


def test_adding_ingredient_to_recipe(client, user, recipe):
    """
    ensure a user can add a ingredient to a recipe
    """
    client.force_authenticate(user)

    ingredient = {
        "quantity": "1",
        "unit": "tablespoon",
        "name": "black pepper",
        "description": "",
    }

    res = client.post(f"/api/v1/recipes/{recipe.id}/ingredients/", ingredient)
    assert res.status_code == status.HTTP_201_CREATED

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK
    assert isinstance(res.json()["sections"], list)

    assert ingredient.get("name") in (
        ingredient.get("name") for ingredient in res.json().get("ingredients")
    ), "ingredient was not in the ingredients of the recipe"


def test_updating_ingredient_of_recipe(client, user, recipe):
    """
    ensure a user can update an ingredient of a recipe
    """
    client.force_authenticate(user)

    ingredient_id = recipe.ingredients[0].id

    ingredient = {"name": "black pepper"}

    res = client.patch(
        f"/api/v1/recipes/{recipe.id}/ingredients/{ingredient_id}/", ingredient
    )
    assert res.status_code == status.HTTP_200_OK

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK

    updated_ingredient = next(
        (x for x in res.json()["ingredients"] if x["id"] == ingredient_id)
    )

    assert updated_ingredient.get("name") == ingredient.get(
        "name"
    ), "ingredient didn't update"


def test_updating_ingredient_position(client, user, recipe):
    """
    ensure we can update the position, necessary for drag and drop in the ui
    """
    client.force_authenticate(user)

    ingredient = recipe.ingredients[0]

    data = {"position": "15"}
    assert data["position"] != ingredient.position

    res = client.patch(
        f"/api/v1/recipes/{recipe.id}/ingredients/{ingredient.id}/", data
    )
    assert res.status_code == status.HTTP_200_OK
    ingredient.refresh_from_db()
    assert res.json()["position"] == data["position"] == ingredient.position


def test_deleting_ingredient_from_recipe(client, user, recipe):
    """
    ensure a user can remove a ingredient from a recipe
    """
    client.force_authenticate(user)

    ingredient_id = recipe.ingredients[0].id

    res = client.delete(f"/api/v1/recipes/{recipe.id}/ingredients/{ingredient_id}/")
    assert res.status_code == status.HTTP_204_NO_CONTENT

    res = client.get(f"/api/v1/recipes/{recipe.id}/")
    assert res.status_code == status.HTTP_200_OK

    assert ingredient_id not in (
        ingredient.get("id") for ingredient in res.json().get("ingredients")
    ), "ingredient was still in the recipe after being deleted"


def test_display_all_accessable_recipes(
    client, user, recipes, team_with_recipes_no_members
):
    """
    User should be able to list all teams they have access to:
        - User-owned recipes
        - Team-owned recipes
    Ensure user can only view team recipes if they have joined.
    """

    client.force_authenticate(user)
    team_with_recipes_no_members.invite_user(user, creator=user)
    res = client.get("/api/v1/recipes/")
    assert res.status_code == status.HTTP_200_OK
    assert len(res.json()) == len(user.recipes.all())

    team_with_recipes_no_members.force_join(user)
    res = client.get("/api/v1/recipes/")
    assert res.status_code == status.HTTP_200_OK
    assert len(res.json()) == len(user.recipes.all()) + len(
        team_with_recipes_no_members.recipes.all()
    )


def test_recording_edits_for_recipes(client, user, recipe):
    """
    ensure edits being recorded for recipes
    """
    client.force_authenticate(user)

    assert recipe.edits == 0

    recipe.name = "A different name"
    recipe.save()

    assert recipe.edits == 1


def test_updating_edit_recipe_via_api(client, user, recipe):
    """
    ensure edits occur when updating the recipe via the api
    """
    client.force_authenticate(user)

    res = client.patch(f"/api/v1/recipes/{recipe.id}/", {"name": "A new name"})
    assert res.status_code == status.HTTP_200_OK

    assert Recipe.objects.get(pk=recipe.id).edits == 1


def test_updating_edit_recipe_via_api_empty_tags(client, user, recipe):
    """
    regression test to allow empty lists for tags
    """
    client.force_authenticate(user)

    res = client.patch(f"/api/v1/recipes/{recipe.id}/", {"tags": []})
    assert res.status_code == status.HTTP_200_OK


def test_copy_recipe(client, user_with_recipes, empty_team, user3):
    """
    Users can copy recipe to team if they have write access.
    """
    recipe = user_with_recipes.recipes.first()
    url = f"/api/v1/recipes/{recipe.id}/copy/"

    # user must own recipe to copy it
    client.force_authenticate(user3)
    assert recipe.owner != user3
    assert (
        client.post(
            url, {"id": empty_team.id, "type": "team", "name": empty_team.name}
        ).status_code
        == status.HTTP_404_NOT_FOUND
    )

    # team viewer cannot add recipe to team
    client.force_authenticate(user_with_recipes)
    empty_team.force_join(user_with_recipes, level=Membership.READ_ONLY)
    assert recipe.owner == user_with_recipes
    assert (
        client.post(url, {"id": empty_team.id, "type": "team"}).status_code
        == status.HTTP_403_FORBIDDEN
    )

    # contributors and admins can add recipe to team
    empty_team.force_join(user_with_recipes, level=Membership.CONTRIBUTOR)
    res = client.post(url, {"id": empty_team.id, "type": "team"})
    assert res.status_code == status.HTTP_200_OK
    assert res.json()["id"] != recipe.id
    assert res.json()["owner"] == {
        "id": empty_team.id,
        "type": "team",
        "name": empty_team.name,
    }


def test_move_recipe(client, user_with_recipes, empty_team, user3):
    """
    Users can move recipe to team if they have write access.
    """
    recipe = user_with_recipes.recipes.first()
    url = f"/api/v1/recipes/{recipe.id}/move/"

    # user must own recipe to copy it
    client.force_authenticate(user3)
    assert recipe.owner != user3
    assert (
        client.post(url, {"id": empty_team.id, "type": "team"}).status_code
        == status.HTTP_404_NOT_FOUND
    )

    # team viewer cannot add recipe to team
    client.force_authenticate(user_with_recipes)
    empty_team.force_join(user_with_recipes, level=Membership.READ_ONLY)
    res = client.post(url, {"id": empty_team.id, "type": "team"})
    assert res.status_code == status.HTTP_403_FORBIDDEN

    # members can add recipe to team
    empty_team.force_join(user_with_recipes, level=Membership.CONTRIBUTOR)
    res = client.post(url, {"id": empty_team.id, "type": "team"})
    assert res.status_code == status.HTTP_200_OK
    assert res.json()["id"] == recipe.id
    assert res.json()["owner"] == {
        "id": empty_team.id,
        "type": "team",
        "name": empty_team.name,
    }


def test_duplicate_recipe_view(
    client: APIClient, user: User, recipe: Recipe, user2: User
) -> None:
    """
    Duplicating a recipe should return the response type of recipe detail
    endpoint
    """

    res = client.post(f"/api/v1/recipes/{recipe.id}/duplicate/")
    assert res.status_code == status.HTTP_403_FORBIDDEN

    client.force_authenticate(user2)

    res = client.post(f"/api/v1/recipes/{recipe.id}/duplicate/")
    assert res.status_code == status.HTTP_404_NOT_FOUND

    client.force_authenticate(user)

    res = client.post(f"/api/v1/recipes/{recipe.id}/duplicate/")
    assert res.status_code == status.HTTP_200_OK

    assert res.json()["id"] != recipe.id
    duped_recipe = Recipe.objects.get(id=res.json()["id"])

    # default clone values for an original
    assert recipe.cloned_at is None
    assert recipe.cloned_by is None
    assert recipe.cloned_from is None

    assert duped_recipe.cloned_by == user
    assert isinstance(duped_recipe.cloned_at, datetime)
    assert duped_recipe.cloned_from == recipe

    assert res.json()["name"] == "Recipe name (copy)"
