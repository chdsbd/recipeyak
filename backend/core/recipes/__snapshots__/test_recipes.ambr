# name: test_recipe_list_timeline_items
  <class 'list'> [
    <class 'dict'> {
      'created': <class 'str'>,
      'created_by': <class 'dict'> {
        'avatar_url': '/avatar/bc6a715808d9aae0ddeefb1e47e482a6?d=identicon&r=g',
        'email': 'john@doe.org',
        'id': <class 'int'>,
      },
      'id': <class 'int'>,
      'last_modified_by': <class 'dict'> {
        'avatar_url': '/avatar/bc6a715808d9aae0ddeefb1e47e482a6?d=identicon&r=g',
        'email': 'john@doe.org',
        'id': <class 'int'>,
      },
      'modified': <class 'str'>,
      'recipe_id': <class 'int'>,
      'text': "I'd say we should use an even smaller pot next time.",
      'type': 'note',
    },
    <class 'dict'> {
      'created': <class 'str'>,
      'created_by': <class 'dict'> {
        'avatar_url': '/avatar/bc6a715808d9aae0ddeefb1e47e482a6?d=identicon&r=g',
        'email': 'john@doe.org',
        'id': <class 'int'>,
      },
      'id': <class 'int'>,
      'modified': <class 'str'>,
      'recipe_id': <class 'int'>,
      'text': 'Use a small pot.',
      'type': 'note',
    },
  ]
---
